### PR TITLE
[benchmark] Add dataUsingUTF8Encoding()

### PR DESCRIPTION
**Pull request by ianpartridge**
Currently there is no microbench for `string.data(using: .utf8)`.

However, there is a microbench which covers creating a `Data` from a `String` via `Data(string.utf8)`.

It is worth having a separate bench for `string.data(using: .utf8)` because it currently bridges via `NSString` whereas `Data(string.utf8)` does not.

Add a benchmark for `string.data(using: .utf8)` so future improvements to its performance can be measured.

CC: @milseman @itaiferber @Catfish-Man

### CODE REVIEW

**Review by parkera — APPROVED** [review]

File: benchmark/single-source/DataBenchmarks.swift [hunk 1]
```diff
@@ -168,17 +168,30 @@ public let DataBenchmarks = [
     runFunction: { string($0*200, from: largeUnicodeData) }, tags: d,
     legacyFactor: 50),
 
-  BenchmarkInfo(name: "StringToDataEmpty",
-    runFunction: { data($0*200, from: emptyString) }, tags: d,
+  BenchmarkInfo(name: "StringToDataFromUTF8ViewEmpty",
```
Comment (itaiferber) [inline benchmark/single-source/DataBenchmarks.swift]: New tests look reasonable to me. I have in the past been requested to not change these unit test names, though. /CC @palimondo for comment

Comment (ianpartridge) [reply benchmark/single-source/DataBenchmarks.swift]: Thanks Itai. We will now have two "StringToData" benches so I thought it was worth naming them, but we could keep the old name if that's required.

Comment (itaiferber) [reply benchmark/single-source/DataBenchmarks.swift]: I completely agree — I think the tests can be renamed. From what I remember, there will be a path forward on renaming these tests while maintaining a mapping to the existing names. The new tests can likely use the new [Benchmark Naming Conventions](https://github.com/apple/swift/blob/master/benchmark/Naming.md) and the old tests will be renamed later (there's some discussion back in #21766 about this)

Comment (palimondo) [reply benchmark/single-source/DataBenchmarks.swift]: Yeah, leave the existing benchmarks unchanged and name the new ones like this:
* `String.data.Empty`
* `String.data.Small`
* `String.data.Medium`

**Review by palimondo — CHANGES_REQUESTED** [review]

File: benchmark/single-source/DataBenchmarks.swift (hunk 1 above)
Comment (palimondo) [reply benchmark/single-source/DataBenchmarks.swift]: There will be another pass in the future to clean up the names in DataBenchmarks to new naming convention, while maintaining long-term performance tracking.

**Comment by ianpartridge** [comment]
@palimondo - I've updated as per feedback.

Please could someone run CI testing?

**Comment by itaiferber** [comment]
@swift-ci Please smoke test

**Review by itaiferber — APPROVED** [review]
Latest changes LGTM

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Review by palimondo — APPROVED** [review]

File: benchmark/single-source/DataBenchmarks.swift [hunk 2]
```diff
@@ -397,12 +410,19 @@ public func string(_ N: Int, from data: Data) {
...
+public func dataFromUTF8View(_ N: Int, from string: String) {
   for _ in 1...N {
     blackHole(Data(string.utf8))
   }
 }
 
+@inline(never)
+public func dataUsingUTF8Encoding(_ N: Int, from string: String) {
+  for _ in 1...N {
+    blackHole(string.data(using: .utf8))
```
Comment (palimondo) [inline benchmark/single-source/DataBenchmarks.swift]: It looks like we are leaking Obj-C objects here… I think we need to do this:
```suggestion
    autoreleasepool { blackHole(string.data(using: .utf8)) }
```

File: benchmark/single-source/DataBenchmarks.swift [hunk 3]
```diff
@@ -169,16 +169,29 @@ public let DataBenchmarks = [
...
+    runFunction: { dataFromUTF8View($0*200, from: mediumString) }, tags: d,
     legacyFactor: 50),
   BenchmarkInfo(name: "StringToDataLargeUnicode",
-    runFunction: { data($0*200, from: largeUnicodeString) }, tags: d,
+    runFunction: { dataFromUTF8View($0*200, from: largeUnicodeString) }, tags: d,
+    legacyFactor: 50),
+
+  BenchmarkInfo(name: "String.data.Empty",
+    runFunction: { dataUsingUTF8Encoding($0*200, from: emptyString) }, tags: d,
+    legacyFactor: 50),
```
Comment (palimondo) [inline benchmark/single-source/DataBenchmarks.swift]: These new benchmarks shouldn't use `legacyFactor`. That's only for getting old benchmarks to run under 1000 μs but report the old inflated runtimes…

As soon as we'll do the renaming of the tests, the runtimes multiplied with legacy factor will be reported with the old names and the real runtimes will be reported under new naming convention (`StringToDataMedium` will become `Data.init.String.Medium`). This way we'll maintain the continuity of long term performance tracking from previous releases until we transition to the fully cleaned up benchmark suite.
```suggestion
  ),
```

**Review by palimondo — CHANGES_REQUESTED** [review]
Two more issues I've overlooked before:
* new tests shouldn't use `legacyFactor`
* benchmark reports shows there's memory leak, we should wrap that conversion in `autoreleasepool`

**Comment by palimondo** [comment]
@milseman @itaiferber Any thoughts on the doubling of memory used between the runs with one and two iterations?
`mem_pages [i1, i2]: min=[3299, 6662]`

Is the use of `@autoreleasepool` required and expected, or is the "leak" a bug that needs to be reported?

**Comment by itaiferber** [comment]
@palimondo This isn't a real leak: the result of `.data(using:)` is an autoreleased `NSData` object. Absent an `autoreleasepool` around the construction of the data, it's going to remain present in memory until the outermost `autoreleasepool` releases it, which in this case is likely past the lifetime of the test.

So wrapping this call up is entirely reasonable

**Comment by ianpartridge** [comment]
@palimondo I have updated this PR as requested.

Please can someone re-run CI and merge as appropriate?

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Review by palimondo — APPROVED** [review]

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please smoke test

**Comment by palimondo** [comment]
@swift-ci please smoke test

**Comment by palimondo** [comment]
@ianpartridge Thank you for the work on these benchmarks!

I'll merge as soon as somebody fixes the failing tests on linux (unrelated to this PR). We can ignore the ⛔️⏱ `String.data.Medium`. I have verified locally that there is no setup overhead... the heuristic can sometimes misfire, when there is high variability between independent runs, as is the case here.

**Comment by ianpartridge** [comment]
Thanks @palimondo, I hope we can merge soon.

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by milseman** [comment]
@swift-ci please smoke test linux platform

**Comment by palimondo** [comment]
@milseman The Linux smoke test has been failing whole weekend. This last one failed few minutes ago. Did somebody fix the CI yet? (it's not even accepting the requests via GitHub without excessive repetition...)

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by milseman** [comment]
Yeah, @gottesmm says he's preparing a fix. Let's wait on that

**Comment by palimondo** [comment]
Oh, I've notices somebody kicked the CI servers and @swift-ci started to respond again, so I've restarted the smoke test hoping that was all it needed… Sorry for wasting the ⚡️🥺

**Comment by milseman** [comment]
https://github.com/apple/swift/pull/22692 I believe

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by palimondo** [comment]
Hmm… smoke test on Linux did successfully finish in #22673. Let's try again…

@swift-ci smoke test linux platform

**Comment by palimondo** [comment]
@milseman Can you please ask around why is the Linux smoke test still failing? This is just an addition of a few benchmarks… the failure is clearly unrelated to this PR! I need to get some 💤😴...

**Comment by milseman** [comment]
The failure is: `TEST 'swift-package-tests :: build-swift-benchmarks-with-swiftpm.txt' FAILED`, which could be related. The error message itself doesn't seem to be so we can try to clean or something. Other PRs are also failing, but for a different reason, so hold on a bit.

**Comment by milseman** [comment]
@swift-ci please clean smoke test linux platform

**Comment by palimondo** [comment]
OK, we are failing to build benchmarks on Linux with swiftpm: `DataBenchmarks.swift:416:5: error: use of unresolved identifier 'autoreleasepool'`…

**Comment by palimondo** [comment]
@swift-ci please smoke test linux platform with #22707

**Comment by palimondo** [comment]
@swift-ci please smoke test linux platform with #22707

**Comment by palimondo** [comment]
Please smoke test with following pull request: 
https://github.com/apple/swift/pull/22707

@swift-ci please smoke test linux platform

**Comment by palimondo** [comment]
For some reason the smoke test with combined PRs didn't work… The error on the console was the same. I guess we'll try again after the #22707 lands?

**Comment by palimondo** [comment]
@swift-ci please smoke test Linux platform

**Comment by palimondo** [comment]
Thank you @ianpartridge!

**Pull request merged**